dandi/dandiapi.py

```diff
@@ -705,16 +705,26 @@ def get_raw_metadata(self) -> Dict[str, Any]:
         else:
             return cast(Dict[str, Any], self.client.get(self.api_path))
 
-    def get_blob_url(self) -> str:
+    def get_content_url(
+        self, regex: str = r".*", follow_redirects: bool = False
+    ) -> str:
         """
-        Returns a non-redirecting URL for downloading the asset, found by
-        inspecting the metadata.  Raises `NotFoundError` if the metadata does
-        not contain an appropriate URL.
+        Returns a URL for downloading the asset, found by inspecting the
+        metadata; specifically, returns the first ``contentUrl`` that matches
+        ``regex``.  If ``follow_redirects`` is true, a ``HEAD`` request is made
+        to resolve any redirects before returning the URL.  Raises
+        `NotFoundError` if the metadata does not contain a matching URL.
         """
         for url in self.get_raw_metadata().get("contentUrl", []):
-            if not url.startswith(self.client.api_url):
-                return url
-        raise NotFoundError("Blob URL not found in asset's contentUrl metadata field")
+            if re.search(regex, url):
+                if follow_redirects:
+                    r = self.client.request("HEAD", url, json_resp=False)
+                    return r.url
```

Comment:
might be a redirect to a redirect: I thought we had a helper to do such jumps until we hit a non-redirecting url but fail to find it -- may be it was removed as part of the girder-related codebase?

anyways -- I feel that we better have a helper for that purpose ;)

seems to be not covered by a test as well

Reply:
`client.request()` calls `session.request()`, which follows all redirects by default.

What exactly do you want a test of this section to do?

Reply:
to follow redirect for the `/api` URL and ensure that no `/api` in the followed up url.

Reply:
Attempting to follow the redirect leads to a 403.  Is that supposed to happen?

Reply:
<details>
<summary>Interesting, confirming with `wget` that for `HEAD` request whenever we reach the S3 we get a 403
</summary> 

```shell
$> wget --method=HEAD -S -O /dev/null https://api.dandiarchive.org/api/dandisets/000027/versions/draft/assets/ff453f4c-a435-4a5d-a48b-128abca5ec47/download/            
Spider mode enabled. Check if remote file exists.
--2021-06-18 16:12:07--  https://api.dandiarchive.org/api/dandisets/000027/versions/draft/assets/ff453f4c-a435-4a5d-a48b-128abca5ec47/download/
Resolving api.dandiarchive.org (api.dandiarchive.org)... 54.162.128.250, 54.157.58.70, 18.205.36.100, ...
Connecting to api.dandiarchive.org (api.dandiarchive.org)|54.162.128.250|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 302 Found
  Connection: keep-alive
  Server: gunicorn/20.0.4
  Date: Fri, 18 Jun 2021 20:12:07 GMT
  Content-Type: text/html; charset=utf-8
  Location: https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826?response-content-disposition=attachment%3B%20filename%3D%22sub-RAT123.nwb%22&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAUBRWC5GAEKH3223E%2F20210618%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210618T201207Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=dd7cc8692557d74e6216133c0923c30f322809283e77cec3e9cc69927698935f
  Vary: Accept, Cookie, Origin
  Allow: GET, HEAD, OPTIONS
  X-Frame-Options: DENY
  Content-Length: 0
  Strict-Transport-Security: max-age=31536000
  X-Content-Type-Options: nosniff
  Referrer-Policy: same-origin
  Via: 1.1 vegur
Location: https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826?response-content-disposition=attachment%3B%20filename%3D%22sub-RAT123.nwb%22&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAUBRWC5GAEKH3223E%2F20210618%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210618T201207Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=dd7cc8692557d74e6216133c0923c30f322809283e77cec3e9cc69927698935f [following]
Spider mode enabled. Check if remote file exists.
--2021-06-18 16:12:07--  https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826?response-content-disposition=attachment%3B%20filename%3D%22sub-RAT123.nwb%22&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAUBRWC5GAEKH3223E%2F20210618%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210618T201207Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=dd7cc8692557d74e6216133c0923c30f322809283e77cec3e9cc69927698935f
Resolving dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)... 52.219.104.84
Connecting to dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)|52.219.104.84|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 403 Forbidden
  x-amz-request-id: VD99N4Z0VPKASPT7
  x-amz-id-2: 7ayWVmtntXgcV0939SU9HGsTNvDT1riiOOIEl3xoTd4EhITSbsqjAQYBP+NTw0Peu5Nc32yTURQ=
  Content-Type: application/xml
  Date: Fri, 18 Jun 2021 20:12:08 GMT
  Server: AmazonS3
Remote file does not exist -- broken link!!!

```
</details>

<details>
<summary>or if I HEAD directly the target URL (works for non-HEAD)</summary> 

```shell
lena:/tmp
$> wget --method=HEAD -S -O /dev/null 'https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826?response-content-disposition=attachment%3B%20filename%3D%22sub-RAT123.nwb%22&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAUBRWC5GAEKH3223E%2F20210618%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210618T201207Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=dd7cc8692557d74e6216133c0923c30f322809283e77cec3e9cc69927698935f'                                                                                                 
Spider mode enabled. Check if remote file exists.
--2021-06-18 16:16:27--  https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826?response-content-disposition=attachment%3B%20filename%3D%22sub-RAT123.nwb%22&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAUBRWC5GAEKH3223E%2F20210618%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210618T201207Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=dd7cc8692557d74e6216133c0923c30f322809283e77cec3e9cc69927698935f
Resolving dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)... 52.219.105.235
Connecting to dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)|52.219.105.235|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 403 Forbidden
  x-amz-request-id: ADCN01KW41N0N6E5
  x-amz-id-2: +bEvBq9YFj7dN48Ym7Zjdrx9FXVss+H6QKxNtbcDEMGGTysIIigCekKKnOeQQCtbU6R5dHTGGyg=
  Content-Type: application/xml
  Date: Fri, 18 Jun 2021 20:16:27 GMT
  Server: AmazonS3
Remote file does not exist -- broken link!!!

(dev3) 1 17047 ->8.....................................:Fri 18 Jun 2021 04:16:28 PM EDT:.
lena:/tmp
$> wget  -S -O /dev/null 'https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826?response-content-disposition=attachment%3B%20filename%3D%22sub-RAT123.nwb%22&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAUBRWC5GAEKH3223E%2F20210618%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210618T201207Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=dd7cc8692557d74e6216133c0923c30f322809283e77cec3e9cc69927698935f'      
--2021-06-18 16:16:34--  https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826?response-content-disposition=attachment%3B%20filename%3D%22sub-RAT123.nwb%22&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAUBRWC5GAEKH3223E%2F20210618%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210618T201207Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=dd7cc8692557d74e6216133c0923c30f322809283e77cec3e9cc69927698935f
Resolving dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)... 52.219.101.187
Connecting to dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)|52.219.101.187|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  x-amz-id-2: HqzdmcQBaI6sLmHV3CdPtLOG4KFKIY7GTtISqqQ5pIXtSq6+4quL1eb43YNaE4+v3juFfb9rc/4=
  x-amz-request-id: X1N029KCJCX9VK94
  Date: Fri, 18 Jun 2021 20:16:35 GMT
  Last-Modified: Wed, 21 Apr 2021 09:38:10 GMT
  ETag: "6ec084ca9d3be17ec194a8f700d65344-1"
  x-amz-version-id: 4BMLdtv04lfi7JDWFfUleHVzYazeFVL4
  Content-Disposition: attachment; filename="sub-RAT123.nwb"
  Accept-Ranges: bytes
  Content-Type: binary/octet-stream
  Server: AmazonS3
  Content-Length: 18792
Length: 18792 (18K) [binary/octet-stream]
Saving to: ‘/dev/null’

/dev/null                                             100%[======================================================================================================================>]  18.35K  --.-KB/s    in 0.05s   

2021-06-18 16:16:34 (380 KB/s) - ‘/dev/null’ saved [18792/18792]

```
</details>


<details>
<summary>but it works if I remove all the "presigning" signature</summary> 

```shell
$> wget --method=HEAD -S -O /dev/null 'https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826'                                                                              
Spider mode enabled. Check if remote file exists.
--2021-06-18 16:17:41--  https://dandiarchive.s3.amazonaws.com/blobs/2db/af0/2dbaf0fd-5003-4a0a-b4c0-bc8cdbdb3826
Resolving dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)... 52.219.96.172
Connecting to dandiarchive.s3.amazonaws.com (dandiarchive.s3.amazonaws.com)|52.219.96.172|:443... connected.
HTTP request sent, awaiting response... 
  HTTP/1.1 200 OK
  x-amz-id-2: BSGGBt1B+Lql75NwX1K8MSS0t4mJHgU54fOZ3+a8AIvzIQIlSlB+TKAeZ44lRqEavbDgD7m/IKM=
  x-amz-request-id: 3JG908CRNANS7BQP
  Date: Fri, 18 Jun 2021 20:17:42 GMT
  Last-Modified: Wed, 21 Apr 2021 09:38:10 GMT
  ETag: "6ec084ca9d3be17ec194a8f700d65344-1"
  x-amz-version-id: 4BMLdtv04lfi7JDWFfUleHVzYazeFVL4
  Accept-Ranges: bytes
  Content-Type: binary/octet-stream
  Server: AmazonS3
  Content-Length: 18792
Length: 18792 (18K) [binary/octet-stream]
Remote file exists.
```
</details>

although soon'ish pre-signed urls would likely to go away (https://github.com/dandi/dandi-api/issues/231#issuecomment-861086394) in the metadata for public assets, we would still have them when we follow the `/download` url.

Before coming up with a "workaround"/alternative, the question to @dchiquito @waxlamp @satra -- is it "by design" that pre-signed urls couldn't be HEAD'ed or there might be some option to enable it, or some bucket setting may be?

Reply:
i believe HEAD/GET requests generate different signatures, and therefore different presigned urls.

Reply:
so it is for dandi-api to discriminate `HEAD` from `GET` for `/download/` end point?

Reply:
i guess it could check the request and pass on the specific when generating the request. the problem is that that url then can be used only for that option.  

i think download could also take an optional parameter `presigned=True|False` that would then either pass the direct url or the presigned one.

Reply:
ok...
- I feel that `follow_redirects` is useful to have but as we see from above -- not really useful ATM.  It is `False` by default, so we could just keep it 
- for @bendichter case I guess it should work without redirects if asking for the `get_content_url('amazon.*/blobs/')` and thus choosing the url (still presigned ATM but AFAIK should be ok) which is hopefully already fully 'redirected' and works for ROS3.

So altogether - this code as good as it would get, and allows to fulfill desired usecase.  Correct @satra @bendichter ?

Reply:
for the moment indeed i think this is fine, but i'll let @bendichter test things out.

Reply:
This looks like the type of thing I would like, but when I test it out locally I'm getting

```python
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-5-d8b7feaf8d24> in <module>
      3         "sub-RAT123/sub-RAT123.nwb"
      4     )
----> 5     url = asset.get_content_url(follow_redirects=True)

~/dev/dandi-cli/dandi/dandiapi.py in get_content_url(self, regex, follow_redirects)
    720                 if follow_redirects:
    721                     r = self.client.request(
--> 722                         "HEAD", url, json_resp=False, allow_redirects=True
    723                     )
    724                     return r.url

~/dev/dandi-cli/dandi/dandiapi.py in request(self, method, path, params, data, files, json, headers, json_resp, retry, **kwargs)
    142             else:
    143                 lgr.error("%s: %s", msg, result.text)
--> 144             raise requests.HTTPError(msg, response=result)
    145 
    146         if json_resp:

HTTPError: Error 403 while sending HEAD request to https://api.dandiarchive.org/api/dandisets/000027/versions/draft/assets/ff453f4c-a435-4a5d-a48b-128abca5ec47/download/
```

Reply:
Yes, S3 doesn't seem to allow HEAD requests.  What exactly should the method do instead for `follow_redirects`?

Reply:
@jwodder This is how I have accomplished this for myself:

```python
s3_url = requests.request(url=url, method='head').url
if '?' in s3_url:
    return s3_url[:s3_url.index('?')]
return s3_url
 ```
 
input: 'https://api.dandiarchive.org/api/dandisets/000054/versions/draft/assets/a3c7865e-3a6a-4f4b-a7cc-fe3c98dbd451/download/'

 output: 'https://dandiarchive.s3.amazonaws.com/blobs/33f/72b/33f72ba7-5ad8-4e42-b52b-e3a6b309aefd'

I don't know much about REST API, so I don't know if I can be much more helpful than that

Reply:
@yarikoptic
* How should this method handle redirects?  Should it just perform a single HEAD request, stop there, and return the new Location, or should it keep going somehow?
* Should query parameters be stripped from the resulting URL?  Should there be an option for that?

Reply:
I guess we could make `follow_redirects` into `Union[bool,int]` so if `int` - it would determine the number of redirects to do, and `True` would follow until no more redirects (or failing like).

I guess we could add `strip_params: bool = False` to mitigate current shortcomings and do what @bendichter did

Reply:
Done.